### ISSUE TITLE
[Bug] Maybe incorrect mathematical formula in `create_softcap_scoremod` (FA4 Score Modulation)

### ISSUE THREAD

**Issue by CaesarG** [Dao-AILab/flash-attention#2396]
### Description

Hi team,

While exploring the new score modulation features in FlashAttention-4, I noticed a mathematical discrepancy in the `create_softcap_scoremod` function located in [`flash_attn/cute/utils.py`](https://github.com/Dao-AILab/flash-attention/blob/main/flash_attn/cute/utils.py). 

The current implementation does not align with the standard Attention Softcapping mechanism (as used in models like Gemma and Grok).

**Current Implementation:**
```python
def create_softcap_scoremod(softcap_val):
    inv_softcap = 1.0 / softcap_val

    @cute.jit
    def scoremod_premask_fn(acc_S_SSA, batch_idx, head_idx, q_idx, kv_idx, aux_tensors):
        scores = acc_S_SSA * inv_softcap
        # BUG: Multiplying by `scores` instead of `softcap_val`
        return scores * cute.math.tanh(scores, fastmath=True) 

    return scoremod_premask_fn
```

**The Issue:**
Mathematically, the current code computes:
`(q * k / C) * tanh(q * k / C)`

Instead of hard-capping the attention scores to the `[-C, C]` range, this formula behaves similarly to a Swish/SiLU activation function. As `q * k` increases, the result grows indefinitely. This defeats the purpose of softcapping, which is to bound the logits and prevent extreme softmax distributions or numerical instability.

**Expected Behavior & Suggested Fix:**
The standard formula for softcapping is:
`C * tanh(q * k / C)`

The function should multiply the `tanh` output by the constant `softcap_val`. Here is the corrected snippet:

```python
def create_softcap_scoremod(softcap_val):
    inv_softcap = 1.0 / softcap_val

    @cute.jit
    def scoremod_premask_fn(acc_S_SSA, batch_idx, head_idx, q_idx, kv_idx, aux_tensors):
        scores = acc_S_SSA * inv_softcap
        # FIXED: Multiply by the constant softcap_val
        return softcap_val * cute.math.tanh(scores, fastmath=True)

    return scoremod_premask_fn
```

Thanks for the amazing work on FA4! Let me know if you would like me to submit a PR for this fix.

**Comment by tridao** [comment]
Are the outputs correct or wrong?

**Comment by CaesarG** [comment]
Hi @tridao,

Thank you for looking into this! I ran some detailed tests, and it appears that the current FA4 output shows a significant numerical deviation compared to the native FA2 and FA3 implementations. However, after making a couple of adjustments, all three versions align perfectly.

The divergence seems to stem from two overlapping details in `flash_attn/cute/utils.py`:

1. **Mathematical adjustment:** The current function calculates `scores * tanh(scores)`. Adjusting it to `softcap_val * tanh(scores)` successfully bounds the logits to the expected `[-C, C]` range.
2. **Parameter signature:** As mentioned in PR #2366, the `seqlen_info` argument is currently missing from `scoremod_premask_fn`, which causes a runtime error when `@cute.jit` passes the arguments.

To verify this, I used the following test script which directly compares FA2, FA3, and FA4 (using `dtype=torch.bfloat16`, `SOFTCAP_VAL=50.0`, and a tolerance of `4e-3`).


```
import torch
from flash_attn import flash_attn_func as flash_attn_func_v2
from flash_attn_interface import flash_attn_func as flash_attn_func_v3
from flash_attn.cute import flash_attn_func as flash_attn_func_v4

def test_softcap_accuracy():
    BATCH = 2
    SEQ_LEN = 256
    HEADS = 128
    HEAD_DIM = 128
    SOFTCAP_VAL = 50.0
    dtype = torch.bfloat16
    device = "cuda"

    print(f"Initializing tensors (Shape: {BATCH}x{SEQ_LEN}x{HEADS}x{HEAD_DIM}, dtype={dtype})...")

    torch.manual_seed(42)
    q = torch.randn(BATCH, SEQ_LEN, HEADS, HEAD_DIM, dtype=dtype, device=device)
    k = torch.randn(BATCH, SEQ_LEN, HEADS, HEAD_DIM, dtype=dtype, device=device)
    v = torch.randn(BATCH, SEQ_LEN, HEADS, HEAD_DIM, dtype=dtype, device=device)

    def get_out(result):
        return result[0] if isinstance(result, tuple) else result

    print("\n[1] Running FlashAttention-2...")
    out_fa2 = get_out(flash_attn_func_v2(q, k, v, softcap=SOFTCAP_VAL, causal=False))

    print("[2] Running FlashAttention-3...")
    out_fa3 = get_out(flash_attn_func_v3(q, k, v, softcap=SOFTCAP_VAL, causal=False))

    print("[3] Running FlashAttention-4...")
    out_fa4 = get_out(flash_attn_func_v4(q, k, v, softcap=SOFTCAP_VAL, causal=False))

    TOLERANCE = 4e-3
    
    def compare_outputs(name, out_a, out_b):
        diff = (out_a - out_b).abs()
        max_err = diff.max().item()
        mean_err = diff.mean().item()
        print(f"\n--- {name} ---")
        print(f"Max Absolute Error: {max_err:.6f}")
        print(f"Mean Absolute Error: {mean_err:.6f}")
        if max_err < TOLERANCE:
            print(f"✅ Test Passed: Outputs match within the tolerance of {TOLERANCE}.")
        else:
            print(f"❌ Test Failed: Significant difference detected! Max error exceeds {TOLERANCE}.")

    print("\n=== Result Comparison ===")
    compare_outputs("FA2 vs FA3", out_fa2, out_fa3)
    compare_outputs("FA2 vs FA4", out_fa2, out_fa4)
    compare_outputs("FA3 vs FA4", out_fa3, out_fa4)

if __name__ == "__main__":
    test_softcap_accuracy()
```

**Before the modification (Original FA4):**


```
--- FA2 vs FA3 ---
Max Absolute Error:  0.003906
Mean Absolute Error: 0.000059
✅ Test Passed: Outputs match within the tolerance of 0.004.

--- FA2 vs FA4 ---
Max Absolute Error:  1.367188
Mean Absolute Error: 0.062988
❌ Test Failed: Significant difference detected! Max error exceeds 0.004.

--- FA3 vs FA4 ---
Max Absolute Error:  1.367188
Mean Absolute Error: 0.062988
❌ Test Failed: Significant difference detected! Max error exceeds 0.004.
```

**The modified implementation:**

```
def create_softcap_scoremod(softcap_val):
    inv_softcap = 1.0 / softcap_val

    @cute.jit
    def scoremod_premask_fn(acc_S_SSA, batch_idx, head_idx, q_idx, kv_idx, seqlen_info, aux_tensors): # Added seqlen_info
        scores = acc_S_SSA * inv_softcap
        return softcap_val * cute.math.tanh(scores, fastmath=True) # Adjusted multiplier

    return scoremod_premask_fn
```

**After the modification (All versions aligned):**


```
--- FA2 vs FA3 ---
Max Absolute Error:  0.003906
Mean Absolute Error: 0.000059
✅ Test Passed: Outputs match within the tolerance of 0.004.

--- FA2 vs FA4 ---
Max Absolute Error:  0.003906
Mean Absolute Error: 0.000052
✅ Test Passed: Outputs match within the tolerance of 0.004.

--- FA3 vs FA4 ---
Max Absolute Error:  0.003906
Mean Absolute Error: 0.000058
✅ Test Passed: Outputs match within the tolerance of 0.004.
```

The outputs match the expected FA2/FA3 behavior beautifully once these changes are applied. I would be very happy to submit a PR incorporating both the signature adjustment and the math formula update if you think this is the proper way to go. Please let me know!

**Comment by CaesarG** [comment]
Hi @tridao, just a quick update: I've gone ahead and opened a PR with the proposed math formula and signature fixes here: #2402 . 

The test results look perfectly aligned now. Please let me know if you need any changes or further testing on this!
